### PR TITLE
Revert "Revert "Update `selenium-webdriver` gem to 4.2.0""

### PR DESCRIPTION
**Pull request by Hamms**
Reverts code-dot-org/code-dot-org#51958, reapplying https://github.com/code-dot-org/code-dot-org/pull/51635. The original attempt consistently failed UI tests on iPhone and iPad when trying to interact with dropdowns, because we were trying to apply a desktop-Safari-specific workaround when executing in mobile Safari.

In Selenium 3, `@browser.browser` is `:Safari` on desktop Safari and `:safari` on mobile. In Selenium 4, it's `:safari` on both, so in addition to browser detection I explicitly inspect browser capabilities (as we do [here](https://github.com/code-dot-org/code-dot-org/blob/2f39e1771d2b67123e8c5f7bfecce9f5b1ba8eff/dashboard/test/ui/features/step_definitions/steps.rb#L1041)) to determine whether or not to apply the workaround.

## Testing Story

Verified locally that the tests pass with the new modifications; also ran with `[test all browsers]` and verified that the only tests which failed in that run did not fail during the actual DTT.

### CODE REVIEW

File: dashboard/test/ui/utils/selenium_browser.rb [hunk 1]
```diff
@@ -41,7 +41,7 @@ def create_response(code, body, content_type)
       if (msg = exception.message.match(/unexpected response, code=(?<code>\d+).*\n(?<error>.*)/))
         error = msg[:error]
         error = JSON.parse(error)['value']['error'] rescue error
-        exception.message.replace("Error #{msg[:code]}: #{error}")
+        raise exception, "Error #{msg[:code]}: #{error}", exception.backtrace
```
Comment (cat5inthecradle) [inline dashboard/test/ui/utils/selenium_browser.rb:44]: I definitely like this approach better than mutating the exception.

**Review by cat5inthecradle — APPROVED** [review]
Testing story looks good!

File: dashboard/test/ui/utils/selenium_browser.rb (hunk 1 above)
Comment (Hamms) [reply dashboard/test/ui/utils/selenium_browser.rb:44]: Likewise! Shout out to @snickell, who actually came up with this before I copied it out of his WIP PR. I assume it was in response to the fact that the message string is now frozen internally by Selenium and throws an error when we try to modify it, but I agree that overall it's much cleaner this way. Also, I learned that you can pass the backtrace manually as an optional third argument to `Kernel#raise`!

**Review by snickell — APPROVED** [review]
Looks good! 

Wish `[test all browsers]` passed on drone :-P, good workaround. For my future ref, here's the drone run for test all browsers: https://drone.cdn-code.org/code-dot-org/code-dot-org/29313/2/5

**Pull request merged**